db/revtree.go

```diff
@@ -301,51 +326,180 @@ func (tree RevTree) copy() RevTree {
-			if gen > maxDeletedLeafGen {
-				maxDeletedLeafGen = gen
+	// Calculate tombstoneGenerationThreshold
+	genShortestNonTSBranch := tree.GenerationShortestNonTombstonedBranch()
+	tombstoneGenerationThreshold := genShortestNonTSBranch - int(maxDepth)
+
+	// Delete nodes whose depth is greater than maxDepth:
+	for revid, node := range tree {
+		if node.depth > maxDepth {
+			delete(tree, revid)
+			pruned++
+		}
+	}
+
+	// Snip dangling Parent links:
+	if pruned > 0 {
+		for _, node := range tree {
+			if node.Parent != "" {
+				if _, found := tree[node.Parent]; !found {
+					node.Parent = ""
+				}
 			}
-		} else if gen > 0 && gen < minLeafGen {
-			minLeafGen = gen
 		}
 	}
 
-	if minLeafGen == math.MaxInt32 {
-		// If there are no non-deleted leaves, use the deepest leaf's generation
-		minLeafGen = maxDeletedLeafGen
+	// Delete any tombstoned branches that are too old
+	leaves := tree.GetLeaves()
+	for _, leafRevId := range leaves {
+		leaf := tree[leafRevId]
+		if !leaf.Deleted { // Ignore non-tombstoned leaves
+			continue
+		}
+		leafGeneration, _ := ParseRevID(leaf.ID)
+		if leafGeneration < tombstoneGenerationThreshold {
+			pruned += tree.DeleteBranch(leaf)
+		}
 	}
-	minGenToKeep := int(minLeafGen) - int(maxDepth) + 1
 
-	if gen := genOfRevID(keepRev); gen > 0 && gen < minGenToKeep {
-		// Make sure keepRev's generation isn't pruned
-		minGenToKeep = gen
+	return
+
+}
+
+func (tree RevTree) DeleteBranch(node *RevInfo) (pruned int) {
+
+	revId := node.ID
+
+	for node := tree[revId]; node != nil; node = tree[node.Parent] {
+		delete(tree, node.ID)
+		pruned++
 	}
 
-	// Delete nodes whose generation is less than minGenToKeep:
-	if minGenToKeep > 1 {
-		for revid, node := range tree {
-			if gen := genOfRevID(revid); gen < minGenToKeep {
-				delete(tree, revid)
-				pruned++
-			} else if gen == minGenToKeep {
-				node.Parent = ""
+	return pruned
+
+}
+
+func (tree RevTree) computeDepths() (maxDepth uint32) {
+
+	// Performance is somewhere between O(n) and O(n^2), depending on the branchiness of the tree.
+	for _, info := range tree {
+		info.depth = math.MaxUint32
+	}
+	// Walk from each leaf to its root, assigning ancestors consecutive depths,
+	// but stopping if we'd increase an already-visited ancestor's depth:
+	for _, revid := range tree.GetLeaves() {
+		var depth uint32 = 1
+		for node := tree[revid]; node != nil; node = tree[node.Parent] {
+			if node.depth <= depth {
+				break // This hierarchy already has a shorter path to another leaf
+			}
+			node.depth = depth
+			if depth > maxDepth {
+				maxDepth = depth
 			}
+			depth++
 		}
 	}
 	return
+
+}
+
+// Find the minimum generation that has a non-deleted leaf.  For example in this rev tree:
+//   http://cbmobile-bucket.s3.amazonaws.com/diagrams/example-sync-gateway-revtrees/three_branches.png
+// The minimim generation that has a non-deleted leaf is "7-non-winning unresolved"
+func (tree RevTree) GenerationShortestNonTombstonedBranch() int {
```

Comment:
The naming of these methods is hard to parse - it's hard to tell whether Generation is supposed to be a noun or a verb. Maybe 'CalculateShortest...' or something?